### DISCUSSION TITLE
Customize special characters

### DISCUSSION THREAD

**Discussion by m-matsubara** [pyscripter/SynEdit#84] [Ideas]
I would like to implement a special character customization feature.

I would like to discuss the spec before creating a pull request.

I'm thinking of creating something like the following as a property, what do you think?

published properties
* SpecialChars: TSynSpecialChars
  * TabGlyph: WideChar
  * TabAlign: TSynSpecialCharsAlign // (sscaLeading, sscaCenter, sscaTrailing)
  * LineBreakGlyph: WideChar
  *  SpaceGlyph: WideChar
  *  IdepgraphicSpaceGlyph: WideChar // White space characters often used in Japanese texts.

Also, I think this is a unique requirement for Japan, but I would like to be able to customize `Idepgraphic space` as a special character independently. Is it okay to include this specification?

**Comment by DGH2112** [comment]
I was looking at the code base to see if this was supportable the other week (I don't think it is at the moment). I receive text files that sometimes contain control or extended characters and wondered if they can be shown similar to Notepad++ and VS Code with little glyths.

**Comment by pyscripter** [comment]
@DGH2112  Please open a different issue on how control characters appear without the eoShowSpecialChars option, showing examples of how they appear in SynEdit and how the appear in VS Code.

**Reply by pyscripter** [reply]
I have opened an issue.  See https://github.com/pyscripter/SynEdit/issues/87.
@m-matsubara @DGH2112 @MShark67 what do you think?

**Reply by pyscripter** [reply]
@DGH2112 This is now committed.  Could  you please try it and report whether it comes close to what you wanted?

**Comment by pyscripter** [comment]
@m-matsubara 
What is the motivation for your proposal.  Is it
- to be able to customize the shown special symbols?
- or just to show a different symbol for the ideographic space?

Also what are the Tabalign characters?

**Comment by pyscripter** [comment]
See also related #89.

**Comment by m-matsubara** [comment]
@pyscripter 
thank you.

I want to customize the Special Characters, especially the line break characters. Currently, in SynEdit, the line break character cannot be changed from `¶` because Special Characters are defined as constants. I would like to customize it to characters such as `↲`, `↓`, and `←`.

TabAlign has wide tabs, so I think it would be nice to be able to change where the glyph is displayed. In my case, I would like to customize the TabGlyph symbol to the left (Leading).

**Reply by pyscripter** [reply]
Happy to change the SynLineBreakGlyph.  I don't like '¶'.  Here is what the U+21B2 looks like:

![image](https://github.com/pyscripter/SynEdit/assets/1311616/16d407b0-7c88-4e93-b424-02946e845500)

Should I commit the change?

Also, I could consider the alignment of the tab symbol.   In fact VS code also aligns it to the left.

I would prefer that to adding further complexity, about something which is does not affect the normal use of the editor.

**Reply by pyscripter** [reply]
Committed the change to SynLineBreakGlyph.   Also I have made the constants in SynEditTypes variables, so that they can be customized.  See https://github.com/pyscripter/SynEdit/commit/b1a00a7bdf525b227cce1bb6df99f9348d07b3f2.

**Reply by m-matsubara** [reply]
Thank you.

I agree with that specification.

I still think `↲` is better for line breaks.

**Reply by pyscripter** [reply]
in https://github.com/pyscripter/SynEdit/commit/c17d42584c91091082b908cee288429533635d2a I have allowed the customization of the alignment of the tab glyph.
